book/src/rfc/rfc-0-generic-parameters.md

```diff
@@ -0,0 +1,359 @@
+
+To ensure no conflicts in the node IDs being regenerated for these instantiated functions, the AST for the main function as an entry point will be regenerated. The monomorphized AST preserves the span information to point to the noname source code for the existing debugging feature.
+
+Same as before, these instantiated functions can be pointed by the expression nodes `ExprKind::FnCall`. With the support of generic parameters, we need to change the way of loading the function AST, as the current fully qualified name pattern doesn't contain the information to differentiate the instantiated functions with different generic values.
+
+Thus we can generate the monomorphized function name, and use it to store the monomorphized function AST instead of the original function name. The new string pattern to store the monomorphized function AST can be:
+`fn_full_qualified_name#generic1=value1#generic2=value2`
+
+
+*Type checking*
+The instantiation of a generic function will resolve the generic types to be concrete types. Similar to the TAST phase, during the monomorphization of a function body, the computed concrete type can be propagated and compared with the return type of the function signature. 
+
+The type check in this phase will always be in concrete type. Any unresolved generic type will fail the type check.
+
+### Monomorphization Process
+Here is an overview of the monomorphization process:
+
+1. Propagates types in the same way the type checker was doing but also with constant values, which will be used to resolve the generic parameters.
+
+2. Along the way, it also propagates the AST nodes. When it is not part of a generic function, the node should remain the same. Otherwise, the node should be regenerated.
+
+3. Whenever it encounters a generic function call, it instantiates the function based on the arguments and store it as a new function with a monomorphized name, then walks through the instantiated function AST. The function call AST node will be modified with the monomorphized name while retaining the same node id and span.
+
+4. In the function instantiation process, all the AST nodes will be regenerated. This new AST will be stored under the monomorphized function name.
+
+5. After monomorphized a function, it should add the name of the original function to a list that records which function AST to delete at the end. We can't not delete the original function AST immediately, because it might be called at different places.
+
+6. In each function block scope, it should type check the return types, by comparing the propagated return type and the defined return type. All these types should be in concrete form without generic parameters involved.
```

Comment:
Thinking more about it I feel like we should be able to typecheck types symbolically and that should be enough… I think my understanding now is that we don’t really need to defer any type checking

Reply:
See above the added list of cases where the type checks can't be done in tast. 
I think it is possible to type check symbolically within a local function scope, since the generic parameters in the scope can remain the same identifiers.
But when it comes to cases that are across function scopes, I am not sure how the equality can be checked as the generic parameters could mean differently in different function scope.